### PR TITLE
[kubeapps-apis] Add direct-helm GetAvailablePackageSummaries

### PR DESCRIPTION
**Pull request by antgamdia**
### Description of the change

This PR just implements the `GetAvailablePackageSummaries` method in the direct-helm plugin.

AFAIK, this method just calls the assetsvc db, no Helm logic is required yet.

~Test pending.~

### Benefits

We will have a brand new `GetAvailablePackageSummaries` operation up and running. 

### Possible drawbacks

N/A

### Applicable issues

  - related #2993
  - depends on https://github.com/kubeapps/kubeapps/pull/3020

### Additional information

~If this approach is good, we should move some files under `pkg`~ No, we shouldn't; we better expose them directly from their current package.

### CODE REVIEW

File: chart/kubeapps/templates/kubeappsapis/deployment.yaml
Comment (absoludity) [inline chart/kubeapps/templates/kubeappsapis/deployment.yaml:102]: We should somehow indicate that this config should be temporary only (ie. we don't want to be providing such config for plugins) until we can switch to the caching provided to plugins that will eventually come out of #2946 . You've said something similar when pulling in the utils.go temporarily below.

**Review by absoludity — COMMENTED** [review]
Sorry, realised this is draft after first comment, so stopped :)

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server_test.go
Comment (absoludity) [inline cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server_test.go:209]: Not sure the k8s client is used currently, but it will be when we add a test ensuring that it 403s for namespaces to which the user doesn't have access, etc.

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server_test.go
Comment (absoludity) [inline cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server_test.go:288]: Currently this is just testing only one case where the namespace is empty? We'd want to ensure that we test the cases:

- specific namespace results a 403 if the user doesn't have access to the ns
- specific namespace results in a query for charts in that ns only otherwise,
- blank namespace results in a query for charts in the kubeapps namespace and any namespace to which the user has access

I think?

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/utils.go
Comment (absoludity) [inline cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/utils.go:33]: Why not ensure this functionality is exported in its current location? If that's an issue, keen to know why. If it's not an issue, I'd recommend just doing it as a separate, pre-quel branch, that is:

-  Public interface `assetManager` -> `AssetManager`
-  Public NewPGManager
-  Public methods

In fact, let me see if I can do that for you to save you the time: #3036 (if you agree, just land that and update).

**Review by absoludity — COMMENTED** [review]
Great work Antonio. I think there's two things missing here:

- existing checks to ensure the user has access to the namespace
- correct handing of a request without the namespace

**Comment by absoludity** [comment]
> Great work Antonio. I think there's two things missing here:
> 
>     * existing checks to ensure the user has access to the namespace
> 
>     * correct handing of a request without the namespace

Actually, thinking more about the last point, perhaps it's enough to just return Unimplemented for now for GetAvailablePackageSummaries without a namespace, given that the current UI doesn't even do this (it was part of ensuring a future possibility with a different workflow). Thoughts?

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go
Comment (antgamdia) [reply cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go:106]: Ooops, just a leftover after copy-n-pasting from the kapp_controller plugin (there, it was initialized to `namespace := globalPackagingNamespace`  instead of  `""`). Thanks for pointing this out

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/utils.go
Comment (antgamdia) [reply cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/utils.go]: Thank you for performing that changes. I initially opted to copy-n-paste the required logic just to have a clear idea of what was really necessary and what's not. 
Also, I didn't want to mix things up, so I just thought, ok let's do it in another PR, but I didn't manage to do it yet, haha.

However, after yours, I've created this another PR https://github.com/kubeapps/kubeapps/pull/3040. More details there, but I had to move the files to a different folder and go package so that I was able to import them. Perhaps I'm missing something.

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go
Comment (antgamdia) [reply cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go:140]: Great, good point here. Thanks for pointing to the auth code (I will comment on my code about it). I've added a check that will prevent users from retrieving packages from ns they can't read.
As you mentioned later, the `ns==""` case won't be implemented yet.

**Comment by antgamdia** [comment]
> Actually, thinking more about the last point, perhaps it's enough to just return Unimplemented for now for GetAvailablePackageSummaries without a namespace, given that the current UI doesn't even do this (it was part of ensuring a future possibility with a different workflow). Thoughts?

That's what I was about to say: let's replicate the current behavior as much as possible, so that we speed up the dev process. Later on, we can start thinking about implementing the TODOs.
I've pushed an initial version of the ns check using the existing logic for the can-i method.

Problem: the client we are passing back to the plugins (via clientGetter) is just for dynamic interfaces, That is, we can't query structured/typed data like a `selfsubjectaccessreviews`. 
Here I've just implemented a temporary workaround, but, my question is: should we also pass to the plugin a normal k8s client besides the existing one?

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go
Comment (antgamdia) [inline cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go:239]: This is what I meant by temporary ugly workaround :P . I just needed to use the dynamic client, so I converted it with `ToUnstructured` and with `FromUnstructured`

File: cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto
Comment (antgamdia) [inline cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:76]: Adding these comments for a better swagger document. 
When embedding these messages inside other msgs we often forgot to add the proper docs, that's why I tend not to embed them. But, on the other side, it's clearer to know which mgs they belong to, so... no strong opinion though.

Comment (absoludity) [reply cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:76]: Yep, and better if we then move the embedded message out. But again, if it's OK with you, I'll add those in the separate PR which hasn't landed yet which adds the filter options. It'll be easier if you don't merge that here nor add functionality depending on them in this PR, but rather keep this one focused on getting the available package summaries (securely, checking the namespace) without filter options. Just so we can work together without depending on unlanded branches.

Comment (absoludity) [reply cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:76]: Or if you want to update once it's landed and still include the filters, that's fine too, but just keen to avoid depending on other people's unlanded code unless it's necessary (ie. if this PR was dependent on filtering options, but I don't think it is).

Comment (absoludity) [reply cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:76]: Hah - so I just found the source of confusion here. Sorry, it was my fault. I'd thought I'd created a PR with the filter options yesterday... but it turns out I'd created the branch before our standup but didn't get back to push it and create the PR. I only found out because I just went to update my branch and pushed the changes to find that it created a new branch on GH. So from your POV you obviously weren't working on something that I'd done at all. Sorry for the confusion :)

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go
Comment (absoludity) [reply cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go:219]: Did this in #3043 . Feel free to land that and update, or just review it and continue your PR without it (and I'll update later).

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go
Comment (absoludity) [reply cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go:239]: Yep, see #3043 .

**Review by absoludity — COMMENTED** [review]
> I've pushed an initial version of the ns check using the existing logic for the can-i method.
> 
> Problem: the client we are passing back to the plugins (via clientGetter) is just for dynamic interfaces, That is, we can't query structured/typed data like a `selfsubjectaccessreviews`.
> Here I've just implemented a temporary workaround, but, my question is: should we also pass to the plugin a normal k8s client besides the existing one?

Yep. I've jumped in to do that separately for you in #3043 .

> 
> btw: tests are still pending (I'll do it tomorrow)

Yep, let me know when it's ready and I'll finish the review. Thanks!

File: cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto
Comment (antgamdia) [reply cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:76]: Hahah, I wasn't getting the point here; I added the FilterOptions in my branch so, as you said, I didn't see you were also working on that. It totally makes sense now.
I'll add the remaining changes to the filter (version and app version) here so that we have the whole filtering functionality implemented.

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go
Comment (antgamdia) [reply cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go:119]: I do agree (I'll change it). However, I decided to leave the `else` statement since, at some point, the `if` branch will have some logic. This way, fewer git changes (aka indentation) when we add the "no namespace" case.
Anyway, happy to remove it.

Comment (absoludity) [reply cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go]: Yep, up to you (I don't mind, just a general note for clarity).

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server_test.go
Comment (absoludity) [inline cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server_test.go:373]: Excellent :)

File: cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go
Comment (absoludity) [inline cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go:222]: Much neater.

**Review by absoludity — APPROVED** [review]
Great, +1. Thanks Antonio. Is it worth implementing the pagination support in the request for this now? What else is required for this method to be used in the client?

File: cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto
Comment (absoludity) [inline cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:406]: I think elsewhere we've stared using `pkg_version` to be specific together with `app_version` as you've used below?

Comment (antgamdia) [reply cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:406]: Good idea indeed. I've added some comments in the doc for unifying the names here. Using just `version` could be misleading in some cases. I'd change each version by either `pkg_` or `app_` prefixes.

**Comment by antgamdia** [comment]
> Is it worth implementing the pagination support in the request for this now? What else is required for this method to be used in the client?

#### Additonal fields for this method

Currently, the client requires the following params  for displaying the catalog: `icon, name, repo, version, description, namespace, id`

A response of this APIs is something like:

```json
{
  "availablePackagesSummaries": [
    {
      "availablePackageRef": {
        "context": {
          "namespace": "kubeapps" namespace
        },
        "identifier": "marketplace/airflow" id
      },
      "latestVersion": "10.2.1", version
      "iconUrl": "https://bitnami.com/assets/stacks/airflow/img/airflow-stack-220x234.png", icon
      "displayName": "airflow", name
      "shortDescription": "Apache Airflow is a platform to programmatically author, schedule and monitor workflows." description
    }
  ]
};
```

So the only missing field is `repo` (it could eventually be retrieved using the `identifier`). 


#### Pagination

Well, this is interesting. Currently, the API is returning something like:

```json
{"data":[ charts ],"meta":{"totalPages": pages }}
```

Perhaps, we could modify the messages to use this `data`/`meta`  approach. Or even go with some of these options:
1) GitHub is using the [`Links` header](https://datatracker.ietf.org/doc/html/rfc5988): https://docs.github.com/en/rest/guides/traversing-with-pagination
2) Google suggests including the pagination as part of the proto msg: https://cloud.google.com/apis/design/design_patterns#list_pagination

However, I'd rather land this PR, continue to work on the other operation and, later on, look into these two additional things in a separate PR. Specially pagination... we should come to an agreement before start changing the UI.
_(I'm gonna crosspost it in the parent issue)_

**Comment by absoludity** [comment]
> However, I'd rather land this PR, continue to work on the other operation and, later on, look into these two additional things in a separate PR. Specially pagination... we should come to an agreement before start changing the UI.

Great, yes - I didn't mean to imply it should be in this PR, just trying to ensure we work on the most relevant things next.

> _(I'm gonna crosspost it in the parent issue)_

Great, I'll reply there :)

**Pull request merged**
